Utils/azuretools-core/src/com/microsoft/azuretools/core/mvp/model/springcloud/AzureSpringCloudMvpModel.java

```diff
@@ -0,0 +1,266 @@
+ * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.microsoft.azuretools.core.mvp.model.springcloud;
+
+import com.microsoft.azure.PagedList;
+import com.microsoft.azure.management.appplatform.v2019_05_01_preview.AppResourceProperties;
+import com.microsoft.azure.management.appplatform.v2019_05_01_preview.DeploymentInstance;
+import com.microsoft.azure.management.appplatform.v2019_05_01_preview.DeploymentResource;
+import com.microsoft.azure.management.appplatform.v2019_05_01_preview.DeploymentResourceProperties;
+import com.microsoft.azure.management.appplatform.v2019_05_01_preview.TestKeys;
+import com.microsoft.azure.management.appplatform.v2019_05_01_preview.implementation.AppPlatformManager;
+import com.microsoft.azure.management.appplatform.v2019_05_01_preview.implementation.AppResourceInner;
+import com.microsoft.azure.management.appplatform.v2019_05_01_preview.implementation.DeploymentResourceInner;
+import com.microsoft.azure.management.appplatform.v2019_05_01_preview.implementation.ServiceResourceInner;
+import com.microsoft.azure.management.resources.Subscription;
+import com.microsoft.azure.storage.StorageException;
+import com.microsoft.azure.storage.file.CloudFile;
+import com.microsoft.azuretools.authmanage.AuthMethodManager;
+import com.microsoft.azuretools.core.mvp.model.AzureMvpModel;
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.HttpException;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import rx.Completable;
+import rx.Observable;
+import rx.schedulers.Schedulers;
+
+public class AzureSpringCloudMvpModel {
+    private static final String LOG_STREAMING_ENDPOINT = "%s/api/logstream/apps/%s/instances/%s?follow=%b";
+
+    public static List<ServiceResourceInner> listAllSpringCloudClusters() throws IOException {
+        final List<ServiceResourceInner> clusters = new ArrayList<>();
+        List<Subscription> subs = AzureMvpModel.getInstance().getSelectedSubscriptions();
+        if (subs.size() == 0) {
+            return clusters;
+        }
+        Observable.from(subs).flatMap((sd) -> Observable.create((subscriber) -> {
+            try {
+                List<ServiceResourceInner> clustersInSubs = listAllSpringCloudClustersBySubscription(
+                        sd.subscriptionId());
+                synchronized (clusters) {
```

Comment:
use `CopyOnWriteArrayList` ?

Reply:
CopyOnWriteArrayList is used in the scenarios where the read operations.